### PR TITLE
kubeadm: Create control plane with ClusterFirstWithHostNet dns policy

### PR DESCRIPTION
**Pull request by andrewrynhard**
**What this PR does / why we need it**:
Currently, the static pods created by `kubeadm` use the `ClusterFirst` dns policy. This prevents doing things like setting the `oidc-issuer-url` flag of the API Server to a Kubernetes service. By using the `ClusterFirstWithHostNet` policy the API Server will be able to resolve the service DNS.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:

**Release note**:

```release-note
kubeadm: create control plane with ClusterFirstWithHostNet DNS policy
```
/cc sig-cluster-lifecycle

### CODE REVIEW

**Comment by neolit123** [comment]
@andrewrynhard 
thanks for the PR!

/kind feature

this might not get reviews for a while due to the 1.12 release closing by.

for reference:
https://godoc.org/k8s.io/kubernetes/pkg/apis/core#DNSPolicy

the addition makes sense, except that we need to verify that there are no unwanted sides effects.

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews 
/cc @rajansandeep @chrisohaver 
for feedback as well.

**Comment by neolit123** [comment]
@andrewrynhard 
please add a release note in the lines of:
```
kubeadm: create control plane with ClusterFirstWithHostNet DNS policy
```
thanks

**Comment by fabriziopandini** [comment]
@andrewrynhard thanks for this PR! I think this is a valuable change
/approve

Being this change on all the control plane components, I left lgtm pending for a second opinion.
/cc @detiber @chuckha @timothysc

**Comment by stealthybox** [comment]
I'm a bit confused on what the current behavior is and why it fails:
```go
    // DNSClusterFirst indicates that the pod should use cluster DNS
    // first unless hostNetwork is true, if it is available, then
    // fall back on the default (as determined by kubelet) DNS settings.
    DNSClusterFirst DNSPolicy = "ClusterFirst"
```
Even if kube-dns is not up yet, won't the default kubelet settings just get used?

Is there an issue that describes this in more detail?

**Comment by andrewrynhard** [comment]
@stealthybox it is a bit confusing. And I'm not sure I understand. I just know that service DNS did not resolve until I switched to self-hosted and that uses `ClusterFirstWithHostNet`. The description in https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/ says
> For Pods running with hostNetwork, you should explicitly set its DNS policy

**Comment by stealthybox** [comment]
Thanks for figuring that out @andrewrynhard 

Looks like this modification is sufficient to cover the etcd, apiserver, KCM, and scheduler:
https://sourcegraph.com/github.com/kubernetes/kubernetes@8f6ec989e006f53c98c107bca17a1d2aeaa82bfa/-/blob/cmd/kubeadm/app/util/staticpod/utils.go#L52:6&tab=references

+1
/lgtm

**Review by stealthybox — APPROVED** [review]

**Comment by BenTheElder** [comment]
The actual functionality for these options is here:

https://github.com/kubernetes/kubernetes/blob/e9fe3f77e9b1810dea6a0918feb1f503ae02fcd7/pkg/kubelet/network/dns/dns.go#L274-L284

I think this broke my funky [docker-in-docker local cluster setup](https://sigs.k8s.io/kind), trying to figure out why :^)

**Comment by BenTheElder** [comment]
AFAICT this is basically equivalent to setting these pods to only use the in-cluster DNS: 
https://github.com/kubernetes/kubernetes/blob/e9fe3f77e9b1810dea6a0918feb1f503ae02fcd7/pkg/kubelet/network/dns/dns.go#L346-L360

Not sure how this is supposed to work for bootstrapping? 
Looks like maybe the API server cannot talk to itself via localhost because the DNS is not up yet.. cc @MrHohn

**Comment by MrHohn** [comment]
> the API Server will be able to resolve the service DNS

Does kubeadm always deploy kube-proxy on master? To be able to send DNS query to kube-dns (coredns), kube-dns's service VIP need to be routable beforehand.

**Comment by BenTheElder** [comment]
I don't think we start kube-proxy during control plane bringup.

**Comment by BenTheElder** [comment]
Issues are attributable back to https://github.com/kubernetes/kubernetes/issues/69195 meaning we cannot resolve `localhost` without DNS in some cases. 😅

**Comment by mauilion** [comment]
The apiserver can't use kube-dns as it's only resolver as it's a circular dependency. The apiserver will come up before the kube-dns pods are present and needs to resolve things like etcd to be able to work.

**Comment by BenTheElder** [comment]
Right. With https://github.com/kubernetes/kubernetes/issues/69195 in some cases we fixed /etc/hosts not being respected as a fallback, which was sufficient for my case. For others this might still be problematic.

**Comment by mauilion** [comment]
it'll probably break external etcd at least

**Comment by mauilion** [comment]
the resolv.conf for when ClusterFirstWithHostNet is set:
```
bash-4.3# cat /etc/resolv.conf 
nameserver 10.96.0.10
search default.svc.cluster.local svc.cluster.local cluster.local cisco.com
options ndots:5
```

This is likely green because the apiserver manifest has the following set:
```
--kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname
--etcd-servers=https://127.0.0.1:2379
```
With these settings we would probably make the tests green cause kube-apiserver will be able to resolve   what it needs to by using the ip addresses. 

If the user has an external etcd cluster this will break.

**Comment by timothysc** [comment]
@mauilion Could you file an issue in the kubeadm repo and we'll assign @chuckha  

priority would be critical.

**Pull request merged**
